File: 16900/BOJ_16975.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    int n;
    long[] bit;

    private void update(int i, long diff) {
        while (i <= n) {
            bit[i] += diff;
            i += i&-i;
        }
    }

    private long query(int i) {
        long sum = 0;
        while (i > 0) {
            sum += bit[i];
            i -= i&-i;
        }
        return sum;
    }

    private void rangeUpdate(int i, int j, long diff) {
        update(i, diff);
        update(j+1, -diff);
    }

    private void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        bit = new long[n+1];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            rangeUpdate(i, i, Integer.parseInt(st.nextToken()));
        }
        int m = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        while (m-->0) {
            st = new StringTokenizer(br.readLine());
            switch (Integer.parseInt(st.nextToken())) {
                case 1: rangeUpdate(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())); break;
                case 2: sb.append(query(Integer.parseInt(st.nextToken()))).append('\n'); break;
            }
        }
        System.out.print(sb);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 16900/BOJ_16975.cpp
#include <iostream>
using namespace std;
#define ll long long

int n;
ll bit[100001] = {0,};

void update(int i, ll diff) {
    while (i <= n) {
        bit[i] += diff;
        i += i&-i;
    }
}

ll query(int i) {
    ll sum = 0;
    while (i > 0) {
        sum += bit[i];
        i -= i&-i;
    }
    return sum;
}

void rangeUpdate(int i, int j, ll diff) {
    update(i, diff);
    update(j+1, -diff);
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int m;
    cin >> n;
    for (int i = 1; i <= n; i++) {
        ll cur;
        cin >> cur;
        rangeUpdate(i, i, cur);
    }

    cin >> m;
    while (m--) {
        int op, a, b, c;
        cin >> op;
        switch (op) {
            case 1:
                cin >> a >> b >> c;
                rangeUpdate(a, b, c);
                break;
            case 2:
                cin >> a;
                cout << query(a) << '\n';
                break;
        }
    }
}
